perf/rwlocks.c

```diff
@@ -0,0 +1,179 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <openssl/pem.h>
+#include <openssl/evp.h>
+#include <openssl/crypto.h>
+#include "perflib/perflib.h"
+
+#define NUM_CALLS_PER_BLOCK         1000
+#define NUM_CALL_BLOCKS_PER_THREAD  1000
+#define NUM_CALLS_PER_THREAD        (NUM_CALLS_PER_BLOCK * NUM_CALL_BLOCKS_PER_THREAD)
+
+int threadcount = 0;
+int err = 0;
+unsigned long *dataval = NULL;
+int writers = 0;
+int readers = 0;
+int write_lock_calls = 0;
+int read_lock_calls = 0;
+OSSL_TIME reader_end = { 0 };
+OSSL_TIME writer_end = { 0 };
+
+CRYPTO_RWLOCK *lock = NULL;
+
+void do_rw_wlock(size_t num)
+{
+    int i;
+    unsigned long *newval, *oldval;
+    int local_write_lock_calls = 0;
+
+    for (i = 0; i < NUM_CALLS_PER_THREAD; i++) {
+        newval = OPENSSL_malloc(sizeof(int));        
+        CRYPTO_THREAD_write_lock(lock);
+        if (dataval == NULL)
+            *newval = 1;
+        else
+            *newval = ((*dataval) + 1);
+        oldval = dataval;
+        dataval = newval;
+        CRYPTO_THREAD_unlock(lock);
+        local_write_lock_calls += 2; /* lock and unlock */
+        OPENSSL_free(oldval);
+    }
+
+    CRYPTO_THREAD_write_lock(lock);
+    write_lock_calls += local_write_lock_calls;
+    writers--;
+    if (writers == 0)
+        writer_end = ossl_time_now();
+    CRYPTO_THREAD_unlock(lock); 
+}
+
+void do_rw_rlock(size_t num)
+{
+    int i;
+    unsigned long last_val = 0;
+    int local_read_lock_calls = 0;
+
+    for (i = 0; i < NUM_CALLS_PER_THREAD; i++) {
+        CRYPTO_THREAD_read_lock(lock);
+        if (dataval != NULL) {
+            if (last_val != 0 && last_val > *dataval)
+                printf("dataval went backwards! %lu:%lu\n", last_val, *dataval);
+            last_val = *dataval;
+        }
+        CRYPTO_THREAD_unlock(lock);
+        local_read_lock_calls += 2; /* lock and unlock */
+    }
+
+    CRYPTO_THREAD_write_lock(lock);
+    read_lock_calls += local_read_lock_calls;
+    readers--;
+    if (readers == 0)
+        reader_end = ossl_time_now();
+    CRYPTO_THREAD_unlock(lock);
+}
+
+void do_rwlocks(size_t num)
+{
+    if (num >= threadcount - writers)
+        do_rw_wlock(num);
+    else
+        do_rw_rlock(num);
+}
+
+int main(int argc, char *argv[])
+{
+    OSSL_TIME duration;
+    OSSL_TIME start;
+    uint64_t us;
+    double avwcalltime;
+    double avrcalltime;
+    int terse = 0;
+    int argnext;
+    char *writeenv;
+
+    if ((argc != 2 && argc != 3)
+       || (argc == 3 && strcmp("--terse", argv[1]) != 0)) {
+        printf("Usage: rwlocks [--terse] threadcount\n");
+        return EXIT_FAILURE;
+    }
+
+    if (argc == 3) {
+        terse = 1;
+        argnext = 2;
+    } else {
+        argnext = 1;
+    }
+
+    threadcount = atoi(argv[argnext]);
+    if (threadcount < 1) {
+        printf("threadcount must be > 0\n");
+        return EXIT_FAILURE;
+    }
+
+    writeenv=getenv("LOCK_WRITERS");
```

Comment:
OpenSSL coding style says to use spaces around the assignment, `writeenv = getenv...`